{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "number :  30\n",
      "name :  MT905416\n",
      "name :  MT890204\n",
      "name :  MT872492\n",
      "name :  MT576584\n",
      "name :  MT549878\n",
      "name :  NC_045512\n",
      "name :  MT876526\n",
      "name :  LC573284\n",
      "name :  MT825091\n",
      "name :  MT820472\n",
      "name :  MT479216\n",
      "name :  MT890462\n",
      "name :  MT704821\n",
      "name :  MT747438\n",
      "name :  MT328032\n",
      "name :  MT966087\n",
      "name :  MT810770\n",
      "name :  MT276328\n",
      "name :  MT263422\n",
      "name :  MT821624\n",
      "name :  MT007544\n",
      "name :  MT612216\n",
      "name :  MT451211\n",
      "name :  MT459987\n",
      "name :  MT706050\n",
      "name :  MT835383\n",
      "name :  MT670021\n",
      "name :  MT907518\n",
      "name :  MT263074\n",
      "name :  MT670014\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## processing: MT905416\n",
      "## processing: MT890204\n",
      "## processing: MT872492\n",
      "## processing: MT576584\n",
      "## processing: MT549878\n",
      "## processing: NC_045512\n",
      "## processing: MT876526\n",
      "## processing: LC573284\n",
      "## processing: MT825091\n",
      "## processing: MT820472\n",
      "## processing: MT479216\n",
      "## processing: MT890462\n",
      "## processing: MT704821\n",
      "## processing: MT747438\n",
      "## processing: MT328032\n",
      "## processing: MT966087\n",
      "## processing: MT810770\n",
      "## processing: MT276328\n",
      "## processing: MT263422\n",
      "## processing: MT821624\n",
      "## processing: MT007544\n",
      "## processing: MT612216\n",
      "## processing: MT451211\n",
      "## processing: MT459987\n",
      "## processing: MT706050\n",
      "## processing: MT835383\n",
      "## processing: MT670021\n",
      "## processing: MT907518\n",
      "## processing: MT263074\n",
      "## processing: MT670014\n",
      "clustalw2 -infile=covid.fasta\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from Bio import Entrez \n",
    "from Bio import SeqIO \n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from Bio import Phylo\n",
    "\n",
    "import pandas as pd \n",
    "\n",
    "def fetch_record(genbank_id): \n",
    "    Entrez.email = \"dnjs0148@naver.com\" \n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=genbank_id, rettype=\"gb\", retmode=\"xml\") \n",
    "    records = Entrez.read(handle) \n",
    "    return records \n",
    "\n",
    "def get_sequence(records): \n",
    "    return records[0][\"GBSeq_sequence\"] \n",
    "\n",
    "#=========================== covid.fasta 파일에 바이러스 염기서열을 모두 써준다. =============================\n",
    "def write_sequence(genbank_id, records): \n",
    "    print(\"## processing: \"+genbank_id) \n",
    "    record_id = records[0][\"GBSeq_accession-version\"] \n",
    "    record_desc = records[0][\"GBSeq_definition\"] \n",
    "    record_seq = records[0][\"GBSeq_sequence\"] \n",
    "    \n",
    "    with open(\"covid.fasta\",'a') as fw: \n",
    "        fw.write(\">\"+record_id+\" \"+record_desc+\"\\n\") \n",
    "        fw.write(record_seq+\"\\n\") \n",
    "        \n",
    "def read_table(file): \n",
    "    df = pd.read_csv(file, sep=\"\\t\") \n",
    "    return df \n",
    "\n",
    "#============================ number만큼 genbank_id를 입력하고 저장한다. ========================\n",
    "f = open(\"covidlist.txt\", 'a')\n",
    "f.write(\"GenBank\"+\"\\t\"+\"RefSeq\"+\"\\t\"+\"Gene Region\"+\"\\n\")\n",
    "\n",
    "a = int(input('number : '))\n",
    "genbank_id = []\n",
    "for b in range(0, a):\n",
    "    genbank_id.append(input('name : '))\n",
    "    \n",
    "for b in range(0, a):\n",
    "    f.write(genbank_id[b]+\"\\t\"+\"\\t\"+\"complete\"+\"\\n\")\n",
    "\n",
    "f.close()\n",
    "\n",
    "#========== txt파일내에 있는 genbank_id에 해당하는 유전자의 염기서열을 covid.fasta 파일에  작성한다. ===========\n",
    "file = \"covidlist.txt\" \n",
    "df = read_table(file) \n",
    "complete_df = df[df[\"Gene Region\"]==\"complete\"] \n",
    "genbank_id_list = complete_df[\"GenBank\"] \n",
    "#genbank_id_list = complete_df[\"GenBank\"] \n",
    "for genbank_id in genbank_id_list: \n",
    "    records = fetch_record(genbank_id) \n",
    "    write_sequence(genbank_id, records)\n",
    "\n",
    "cline = ClustalwCommandline(\"clustalw2\", infile=\"covid.fasta\")\n",
    "print(cline)\n",
    "\n",
    "# clustalw_exe = r\"C:\\Users\\dnjs0\\clustalw2.exe\"                                          필요할 때 주석처리 없애고 본인꺼로 바꿔서 ㄱㄱ\n",
    "clustalw_exe = r\"C:\\Program Files (x86)\\ClustalW2\\clustalw2.exe\"\n",
    "clustalw_cline = ClustalwCommandline(clustalw_exe, infile=\"covid.fasta\")\n",
    "assert os.path.isfile(clustalw_exe), \"Clustal W executable missing\"\n",
    "\n",
    "stdout, stderr = clustalw_cline()\n",
    "\n",
    "f = open('covid.aln', 'r')\n",
    "while True:\n",
    "    line = f.readline()\n",
    "    if not line: break\n",
    "    print(line)\n",
    "\n",
    "tree = Phylo.read(\"covid.dnd\", \"newick\")\n",
    "Phylo.draw_ascii(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
